lib/datadog/grok/src/parse_grok_rules.rs

```diff
@@ -0,0 +1,375 @@
+
+You can write parsing rules with the %{MATCHER:EXTRACT:FILTER} syntax:
+- Matcher: A rule (possibly a reference to another token rule) that describes what to expect (number, word, notSpace, etc.)
+- Extract (optional): An identifier representing the capture destination for the piece of text matched by the Matcher.
+- Filter (optional): A post-processor of the match to transform it.
+
+Each rule can reference parsing rules defined above itself in the list.
+Only one can match any given log. The first one that matches, from top to bottom, is the one that does the parsing.
+For further documentation and the full list of available matcher and filters check out https://docs.datadoghq.com/logs/processing/parsing
+*/
+pub fn parse_grok_rules(
+    helper_rules: &[String],
+    parsing_rules: &[String],
+) -> Result<Vec<GrokRule>, Error> {
+    let mut prev_rule_definitions: HashMap<&str, String> = HashMap::new();
+    let mut prev_rule_filters: HashMap<&str, HashMap<LookupBuf, Vec<GrokFilter>>> = HashMap::new();
+    let mut grok = initialize_grok();
+
+    // parse helper rules to reference them later in the match rules
+    parse_rules(
+        helper_rules,
+        &mut prev_rule_definitions,
+        &mut prev_rule_filters,
+        &mut grok,
+    )?;
+    // parse match rules and return them
+    parse_rules(
+        parsing_rules,
+        &mut prev_rule_definitions,
+        &mut prev_rule_filters,
+        &mut grok,
+    )
+}
+
+fn parse_rules<'a>(
+    parsing_rules: &'a [String],
+    mut prev_rule_definitions: &mut HashMap<&'a str, String>,
+    mut prev_rule_filters: &mut HashMap<&'a str, HashMap<LookupBuf, Vec<GrokFilter>>>,
+    mut grok: &mut Grok,
+) -> Result<Vec<GrokRule>, Error> {
+    parsing_rules
+        .iter()
+        .filter(|&r| !r.is_empty())
+        .map(|r| {
+            parse_grok_rule(
+                r,
+                &mut prev_rule_definitions,
+                &mut prev_rule_filters,
+                &mut grok,
+            )
+        })
+        .collect::<Result<Vec<GrokRule>, Error>>()
+}
+
+fn parse_grok_rule<'a>(
+    rule: &'a str,
+    mut prev_rule_definitions: &mut HashMap<&'a str, String>,
+    mut prev_rule_filters: &mut HashMap<&'a str, HashMap<LookupBuf, Vec<GrokFilter>>>,
+    grok: &mut Grok,
+) -> Result<GrokRule, Error> {
+    let mut split_whitespace = rule.splitn(2, ' ');
+    let split = split_whitespace.by_ref();
+    let rule_name = split.next().ok_or_else(|| {
+        Error::InvalidGrokExpression(
+            rule.to_string(),
+            "format must be: 'ruleName definition'".into(),
+        )
+    })?;
+    let mut rule_def = split
+        .next()
+        .ok_or_else(|| {
+            Error::InvalidGrokExpression(
+                rule.to_string(),
+                "format must be: 'ruleName definition'".into(),
+            )
+        })?
+        .to_string();
+
+    let rule_def_cloned = rule_def.clone();
```

Comment:
I believe this clone is unnecessary as when it is being used later it is being used with `as_str`, so only a reference to the original value should be needed without consuming it.